src/com/ms/silverking/cloud/dht/daemon/storage/FileSegment.java

```diff
@@ -23,243 +23,255 @@
-        dataBuf = raFile.getChannel().map(MapMode.READ_ONLY, 0, dataSegmentSize);
-        raFile.close();
-        return dataBuf;
-    }
-    
-    static final FileSegment openForRecovery(File nsDir, int segmentNumber, int dataSegmentSize, 
-                                            SyncMode syncMode, NamespaceOptions nsOptions) throws IOException {
-        RandomAccessFile    raFile;
-        ByteBuffer          dataBuf;
-        FileSegment         segment;
-        
-        raFile = new RandomAccessFile(fileForSegment(nsDir, segmentNumber), syncModeToFileOpenMode(syncMode));
-        
-        dataBuf = raFile.getChannel().map(MapMode.READ_WRITE, 0, dataSegmentSize);
-        segment = new FileSegment(nsDir, segmentNumber, AccessMode.Repair, raFile, dataBuf, dataSegmentSize, nsOptions);
-        return segment;
+      } else {
+        dataBuf = null;
+      }
     }
-    
-    // Open read-only or for update
-    private FileSegment(File nsDir, int segmentNumber, AccessMode accessMode, RandomAccessFile raFile, 
-                    ByteBuffer dataBuf, CuckooBase keyToOffset, Set<Integer> invalidatedOffsets, 
-                    OffsetListStore offsetListStore, int dataSegmentSize, NamespaceOptions nsOptions) 
-                    throws IOException {
-        super(nsDir, segmentNumber, dataBuf, keyToOffset, invalidatedOffsets, offsetListStore, dataSegmentSize);
-        this.accessMode = accessMode;
-        if (accessMode != AccessMode.ReadOnly && accessMode != AccessMode.Update && accessMode != AccessMode.ReadIndexOnly) {
-            throw new RuntimeException("Unexpected mode: "+ accessMode);
-        }
-        raFile.close();
-        this.raFile = null;
-        this.storageFormat = FileSegmentStorageFormat.parse(nsOptions.getStorageFormat());
+    if (debugMetaData) {
+      Log.warningf("dataSegmentSize %d", dataSegmentSize);
+      Log.warningf("raFile.length() %d", raFile.length());
+      Log.warningf("raFile.length() - dataSegmentSize %d", raFile.length() - dataSegmentSize);
     }
+    if (segmentIndexLocation == SegmentIndexLocation.RAM) {
+      byte[] _htBufArray;
 
-    // called from Create
-    private FileSegment(File nsDir, int segmentNumber, AccessMode accessMode, RandomAccessFile raFile, 
-            ByteBuffer dataBuf, int dataSegmentSize, NamespaceOptions nsOptions) 
-                    throws IOException {
-        super(nsDir, segmentNumber, dataBuf, StoreConfiguration.fileInitialCuckooConfig, dataSegmentSize, nsOptions);
-        this.accessMode = accessMode;
-        this.raFile = raFile;
-        this.storageFormat = FileSegmentStorageFormat.parse(nsOptions.getStorageFormat());
+      _htBufArray = new byte[(int) (raFile.length() - dataSegmentSize)];
+      raFile.seek(dataSegmentSize);
+      raFile.read(_htBufArray);
+      rawHTBuf = ByteBuffer.wrap(_htBufArray);
+    } else {
+      rawHTBuf = raFile.getChannel().map(MapMode.READ_ONLY, dataSegmentSize, raFile.length() - dataSegmentSize);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *getChannel*. Currently, there are execution paths that do not contain closure statements, for example, when FileChannel.map throws an exception. Either a) close the object returned by *getChannel()* in a try-finally block or b) close the resource by declaring the object returned by *getChannel()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).